### DISCUSSION TITLE
What's the relationship between "compile_commands.json", project configuration file(.clangd file) and clangd command line args?

### DISCUSSION THREAD

**Discussion by DFVSQY** [clangd/clangd#1335] [Q&A] [answered]
I have read clangd documentation and have a try it in vscode, but i was a bit confused about  the relationship among `compile_commands.json`, `.clangd file` and `clangd command line args`. 

It seems all configurations will affect clangd, for example, I can open `-Wall` compile option in these three places.  What is difference among them? Is there a priority for clangd?

**Comment by HighCommander4** [comment, marked as answer]
It's important to distinguish between two types of configuration info:

 1. Flags to the compiler commands used to process your code. This is the category that `-Wall` belongs to.
 2. Flags passed to clangd itself that affect its features, e.g. "whether code completion for a function call should insert placeholders for the function arguments".

`compile_commands.json` is the standard way to specify (1) for all clang-based tools, including clangd. The file is typically generated by the project's build system, thereby allowing information (such as the location of include paths) to be specified in only one place (the build system's metadata files), and have things be consistent between the actual build and what clangd sees.

Clangd also allows modifying (1) using the clangd config file's [CompileFlags](https://clangd.llvm.org/config.html#compileflags) block, in the form of adding and removing flags. These are "edits" applied to the command taken form `compile_commands.json`, and they're meant for situations where for some reason you want clangd to understand your code differently than your build.

Flags in category (2) are specified using a mixture of clangd command-line args, and the clangd config file. Generally, we have been trying to move away from command-line args and towards the config file, so most newer options are specified via the config file. However, there are some cases the command-line is still more appropriate (e.g. since the command line is usually specified in a client-specific place (e.g. vscode's `"clangd.arguments"`), options that you may only want to use with a particular client make more sense there).
